internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -236,7 +237,7 @@ func (r *Reconciler) computeDesiredMachineSet(deployment *clusterv1.MachineDeplo
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to compute desired MachineSet: failed to compute machine template hash")
 		}
-		uniqueIdentifierLabelValue = fmt.Sprintf("%d", templateHash)
+		uniqueIdentifierLabelValue = fmt.Sprintf("%d-%s", templateHash, apirand.String(5))
```

Comment:
We need this to make sure that the second MS (created as a result of rolloutAfter) is different from the first MS (the one existing before rolloutAfter). If not, we will end up "creating" a duplicate MS that is an exact copy of the existing MS. Because we use SSA we end up updating the existing MS instead of creating a new one which is not desired.

With out this change the old MS and the new MS are the same MS and the reconciler will try to continuously scale up and and scale down the same MS indefinitely. Also, since we ended up updating the same MS instead of creating a new one its creationTimestamp is not changed therefore the MS will always satisfy the "shouldRolloutAfter" criteria and will keep rolling out forever.


This change will fix all these problems. Because of this change we will end up creating a new MS as desired and a successful rollout will be performed.

Reply:
Makes sense, nit: it'd be nice to clarify that with comment for uniqueIdentifierLabelValue